Fixing broken bind mount on CentOS 7

jtyr opened the pull request:
##### ISSUE TYPE

Bugfix Pull Request

##### COMPONENT NAME

`mount` module

##### ANSIBLE VERSION

```
ansible 2.2.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY

I have noticed that when I have bind mounts on CentOS7, the module doesn't detect that the mount point is already mounted and mounts it again. The reason behind that was that the source path was wrongly dedicted from the `mountinfo` file. This PR is simplifying the way how we detect the source by following the `parent_id` to the parent `id`. Tested on CentOS 6 and 7 and Ubuntu 12 and 14.

jtyr committed:
Fixing broken bind mount on CentOS 7

jtyr commented:
@abadger Please review.

abadger commented on lib/ansible/modules/system/mount.py:
```diff
@@ -20,9 +20,20 @@
+from ansible.module_utils.pycompat24 import get_exception
+from ansible.module_utils.six import iteritems
+from ansible.module_utils._text import to_native
+import os
```
As long as you're doing the work in this file, I don't care about the rest of the style changes but the imports need to go back where you found them.  We've told you several times that the boilerplate ansible variables at the top of the modules are not used by the module code and therefore they should be treated the same as the other boilerplate headers to code.  From the module's POV, these are comments. They use python variables so that other code can easily pull them out of the file using the ast module.

(This is written down in the "New Files" section of https://github.com/ansible/proposals/issues/50 )

abadger commented:
The changes that affect the execution look okay but I'm not an expert on the various formats that are being parsed and what needs to be done with them.  Could you add an integration test that failed on centos7 before and works once this is added?  That will give greater confidence, not just about this code, but also about future changes to the module in the future.

abadger commented:
@jtyr, 2.3.0 and 2.2.2 are somewhere in the not too distant future.  Any progress on an integration test and moving the imports?

jtyr commented:
@abadger I will into it tomorrow.

jtyr commented:
@abadger I had a look into the integration test you already have (`test/integration/targets/mount/tasks/main.yml`) and you are already testing for double mounting. I have tested it in Vagrant with the following playbook:

```
---

- name: Testing mount module
  hosts: all
  become: yes
  vars:
    mnt_src: /tmp/a/b/c/aaa
    mnt_dst: /tmp/a/b/c/bbb
  tasks:
    - name: Create mount points
      file:
        path: "{{ item }}"
        state: directory
      with_items:
        - "{{ mnt_src }}"
        - "{{ mnt_dst }}"

    - name: Mount bind 1
      mount:
        name: "{{ mnt_dst }}"
        src: "{{ mnt_src }}"
        opts: bind
        fstype: None
        state: mounted

    - name: Mount bind 2
      mount:
        name: "{{ mnt_dst }}"
        src: "{{ mnt_src }}"
        opts: bind
        fstype: None
        state: mounted

    - name: Unmount bind 1
      mount:
        name: "{{ mnt_dst }}"
        state: unmounted

    - name: Unmount bind 2
      mount:
        name: "{{ mnt_dst }}"
        state: unmounted
```

And this is the result with the current `mount` module:

```
PLAY [Testing mount module] ****************************************************

TASK [setup] *******************************************************************
ok: [test-mount]

TASK [Create mount points] *****************************************************
changed: [test-mount] => (item=/tmp/a/b/c/aaa)
changed: [test-mount] => (item=/tmp/a/b/c/bbb)

TASK [Mount bind 1] ************************************************************
changed: [test-mount]

TASK [Mount bind 2] ************************************************************
changed: [test-mount]

TASK [Unmount bind 1] **********************************************************
changed: [test-mount]

TASK [Unmount bind 2] **********************************************************
changed: [test-mount]

PLAY RECAP *********************************************************************
test-mount                 : ok=6    changed=5    unreachable=0    failed=0
```

And this is the result when I run it wit the `mount` module from this PR:

```
PLAY [Testing mount module] ****************************************************

TASK [setup] *******************************************************************
ok: [test-mount]

TASK [Create mount points] *****************************************************
ok: [test-mount] => (item=/tmp/a/b/c/aaa)
ok: [test-mount] => (item=/tmp/a/b/c/bbb)

TASK [Mount bind 1] ************************************************************
changed: [test-mount]

TASK [Mount bind 2] ************************************************************
ok: [test-mount]

TASK [Unmount bind 1] **********************************************************
changed: [test-mount]

TASK [Unmount bind 2] **********************************************************
ok: [test-mount]

PLAY RECAP *********************************************************************
test-mount                 : ok=6    changed=2    unreachable=0    failed=0
```

As you can see, the current `mount` module mounts it twice where the module from this PR only once. The reason for this could be that the mounting doesn't work the same way in Docker and in VM.

jtyr commented:
I was looking into the remount problem today and I think that the problem is that in the `mount()` function we test if it's already mounted with the `ismount()` function which doesn't handle bind mounts. We must put there similar logic we use in the `main()` function to handle the bind mounts separately. I will try to fix it over the weekend or on Monday.

jtyr committed:
Fixing remount

jtyr committed:
Moving imports to the 'right' place

abadger commented on lib/ansible/modules/system/mount.py:
```diff
@@ -131,14 +137,14 @@
     state: present
 '''
 
-import os
```
import os belongs here rather than mixed in with the ansible imports because it's from the stdlib (group 1 in pep8) while the ansible imports are application specific imports (group 3)

jtyr replied:
Fixed.

jtyr committed:
Fixing PEP8 compliance

jtyr committed:
Reordering imports - stdlib first
